server/src/main/java/org/activityinfo/server/command/handler/TransferDatabaseHandler.java

```diff
@@ -0,0 +1,94 @@
+package org.activityinfo.server.command.handler;
+
+import com.google.inject.Inject;
+import org.activityinfo.legacy.shared.AuthenticatedUser;
+import org.activityinfo.legacy.shared.command.TransferDatabase;
+import org.activityinfo.legacy.shared.command.result.CommandResult;
+import org.activityinfo.legacy.shared.command.result.TransferResult;
+import org.activityinfo.server.database.hibernate.entity.Database;
+import org.activityinfo.server.database.hibernate.entity.User;
+import org.activityinfo.server.mail.MailSender;
+import org.activityinfo.server.mail.RejectDatabaseTransferMessage;
+import org.activityinfo.server.mail.SuccessfulDatabaseTransferMessage;
+
+import javax.persistence.EntityManager;
+
+public class TransferDatabaseHandler implements CommandHandler<TransferDatabase> {
+
+    private final EntityManager entityManager;
+    private final MailSender mailSender;
+    private User currentUser;
+
+    @Inject
+    public TransferDatabaseHandler(EntityManager entityManager, MailSender mailSender) {
+        this.entityManager = entityManager;
+        this.mailSender = mailSender;
+    }
+
+    @Override
+    public CommandResult execute(TransferDatabase cmd, User currentUser) {
+        this.currentUser = currentUser;
+
+        Database database = getDatabase(cmd.getDatabaseId());
+        User requester = database.getOwner();
+        User approver = database.getTransferUser();
+        checkPreconditions(cmd, requester, approver, database);
+
+        if (cmd.getDecision() == TransferDatabase.Decision.APPROVE) {
+            transferDatabase(database, approver);
+            clearTransferToken(database);
+            sendSuccessNotifications(requester, approver, database);
+        } else {
+            clearTransferToken(database);
+            sendRejectionNotification(requester, approver, database);
+        }
+
+        TransferResult result = new TransferResult();
+        result.setResult(cmd.getDecision() == TransferDatabase.Decision.APPROVE
+                ? TransferResult.Result.TRANSFER_SUCCESSFUL
+                : TransferResult.Result.TRANSFER_REJECTED);
+
+        return result;
+    }
+
+    private void sendRejectionNotification(User originalOwner, User newOwner, Database database) {
+        mailSender.send(new RejectDatabaseTransferMessage(originalOwner, newOwner, database));
+    }
+
+    private void sendSuccessNotifications(User originalOwner, User newOwner, Database database) {
+        mailSender.send(new SuccessfulDatabaseTransferMessage(originalOwner, database));
+        mailSender.send(new SuccessfulDatabaseTransferMessage(newOwner, database));
+    }
+
+    private void clearTransferToken(Database database) {
+        database.setTransferToken(null);
+        database.setTransferUser(null);
+        database.setTransferRequestDate(null);
+        entityManager.persist(database);
+    }
+
+    private void transferDatabase(Database database, User approver) {
+        database.setOwner(approver);
```

Comment:
What if the new owner has a UserPermission record in this database? Shouldn't that be removed?
Should the former owner be given design privileges by default so they don't completely loose access?

Reply:
Will add to next iteration. 

Regarding access for former owners, I'm not sure. What is the common case for transferring ownership? If it is handover management because users are moving to another project, then they shouldn't retain that level of permission. If its simple transfer between supervisors, maybe they should maintain that level of rights.

I think the safest case here might be to default to only View permissions, and let the new owner decide the access level (or choose to remove them). What do you think?